### PR TITLE
Setting: Enforce constructor signature

### PR DESCRIPTION
**Pull request by jayvdb**
Fixes https://github.com/coala/coala/issues/4202

### CODE REVIEW

**Comment by jayvdb** [comment]
There is a slight breaking change in here, that a combination of invalid `key` and `origin` parameter values could previously raise a `ValueError` in the `key.setter` will now be a TypeError.

IMO that is too niche a scenario to support the old exception behaviour.

**Comment by adtac** [comment]
```bash
$ grep "Setting([^,]*,[^,]*," -r -n **/*.py -C 1
coalib/bearlib/languages/LanguageDefinition.py:89:            value = Setting(item, ', '.join(escape(val, ',') for val in value))
coalib/bearlib/languages/LanguageDefinition.py:91:            value = Setting(item, ', '.join(
coalib/settings/ConfigurationGathering.py:271:    >>> section.append(Setting("files", "**", origin="/tmp/.coafile"))
coalib/settings/ConfigurationGathering.py:287:    >>> section.append(Setting('project_dir', os.path.abspath('/tmp'), '/'))
tests/bearlib/abstractions/SectionCreatableTest.py:54:        section.append(Setting('setting_three', '2, 4'))
tests/bearlib/abstractions/SectionCreatableTest.py:55:        section.append(Setting('opt_raw_set', 'tst ,'))
* tests/bears/BearTest.py:322:        section.append(Setting('files', '**', '/path/to/dir/config'))
* tests/core/BearTest.py:175:        section.append(Setting('files', '**', '/path/to/dir/config'))
+ tests/settings/SectionTest.py:21:        uut.append(Setting(5, 5, 5))
+ tests/settings/SectionTest.py:39:        uut.append(Setting(5, 5, 5))
+ tests/settings/SectionTest.py:40:        uut.add_or_create_setting(Setting('TEsT', 4, 5))
+ tests/settings/SectionTest.py:41:        defaults.append(Setting('tEsT', 1, 3))
+ tests/settings/SectionTest.py:42:        defaults.append(Setting(' great   ', 3, 8))
+ tests/settings/SectionTest.py:43:        defaults.append(Setting(' great   ', 3, 8), custom_key='custom')
+ tests/settings/SectionTest.py:44:        uut.add_or_create_setting(Setting('custom', 4, 8, to_append=True))
+ tests/settings/SectionTest.py:45:        uut.add_or_create_setting(Setting(' NEW   ', 'val', 8))
+ tests/settings/SectionTest.py:46:        uut.add_or_create_setting(Setting(' NEW   ', 'vl', 8),
- tests/settings/SectionTest.py:48:        uut.add_or_create_setting(Setting('new', 'val', 9),
n tests/settings/SectionTest.py:179:        section.append(Setting('bear_dirs', 'test1, test2 (1)'))
* tests/settings/SettingTest.py:30:        self.uut = Setting('key', ' 22', '')
- tests/settings/SettingTest.py:37:        self.uut = Setting('key', '.',
- tests/settings/SettingTest.py:45:        self.uut = Setting('key', '., ' + abspath.replace('\\', '\\\\'),
- tests/settings/SettingTest.py:61:        self.uut = Setting('key', '., ' + abspath.replace('\\', '\\\\'),
- tests/settings/SettingTest.py:69:        self.uut = Setting('key', '1, 2, 3')
- tests/settings/SettingTest.py:74:            self.uut = Setting('key', '1, a, 3')
- tests/settings/SettingTest.py:78:        self.uut = Setting('key', '1, 2: t, 3')
- tests/settings/SettingTest.py:83:            self.uut = Setting('key', '1, a, 3')
- tests/settings/SettingTest.py:87:        self.uut = Setting('key', '1, 2: t, 3')
- tests/settings/SettingTest.py:92:            self.uut = Setting('key', '1, a, 3')
* tests/settings/SettingTest.py:96:        self.uut = Setting('key', ' 22\n', '.', strip_whitespaces=True)
* tests/settings/SettingTest.py:103:            self.uut = Setting('key', '22\n', '.', to_append=True)
* tests/settings/SettingTest.py:108:            self.uut = Setting('key', '1, 2, 3', '.', to_append=True)
* tests/testing/LocalBearTestHelperTest.py:25:        section.append(Setting('result', 'a, b'))
```

Legend:

`+`: you've taken care of these lines
`-`: these are invalid because I'm too lazy to write a better regex
`*`: you've probably missed these ;)

File: coalib/settings/Setting.py [hunk 1]
```diff
@@ -93,15 +96,16 @@ class Setting(StringConverter):
...
 
+    @enforce_signature
     def __init__(self,
                  key,
                  value,
-                 origin='',
-                 strip_whitespaces=True,
+                 origin: str='',
+                 strip_whitespaces: bool=True,
                  list_delimiters=(',', ';'),
```
Comment (adtac) [inline coalib/settings/Setting.py]: `list_delimiters: tuple=(',', ';')`?

File: coalib/settings/Setting.py [hunk 2]
```diff
@@ -93,15 +96,16 @@ class Setting(StringConverter):
     conversions into common data types.
     """
 
+    @enforce_signature
     def __init__(self,
                  key,
```
Comment (adtac) [inline coalib/settings/Setting.py]: let's enforce this to be `str` as well ;) who knows what'll happen in the future :D

Comment (adtac) [inline coalib/settings/Setting.py]: nvm it's typecast as a str in line 144

**Comment by jayvdb** [comment]
@adtac, I dont get your meaning wrt the regex at https://github.com/coala/coala/pull/4203#issuecomment-300064174 ; the ones you say I have probably missed are all `origin: str` , and thus not particularly interesting for this change.

File: coalib/settings/Setting.py (hunk 2 above)
Comment (jayvdb) [inline coalib/settings/Setting.py]: Ya, many breakages occur if it is enforced to be `str`.

**Comment by adtac** [comment]
ohhhhh nvm then :D

**Comment by adtac** [comment]
ack https://github.com/coala/coala/pull/4203/commits/98a36066eb0c1c8c77cd1bfa58db2b73b89747b9

File: coalib/settings/Setting.py (hunk 1 above)
Comment (jayvdb) [inline coalib/settings/Setting.py]: StringConverter.list_delimiters only requires it is a Iterable

Comment (sils) [inline coalib/settings/Setting.py]: well it could be any reusable iterable, at least tuple list and set are ok IIRC

**Comment by sils** [comment]
ack https://github.com/coala/coala/pull/4203/commits/2065f4c73cdce96d13b4962f3e3cf1f7be21825f

**Comment by sils** [comment]
ack https://github.com/coala/coala/pull/4203/commits/98a36066eb0c1c8c77cd1bfa58db2b73b89747b9

**Comment by jayvdb** [comment]
@rultor merge

**Pull request merged**